### PR TITLE
feat(playground): accumulate errors while parsing span attributes

### LINKED ISSUE #4922: [playground] accumulate errors while parsing span attributes for prompt playground
Accumulate errors while trying to parse span attributes to pre-populate the playground for span replay.

Attempt to parse span attributes starting with the most important and most strict parsing, accumulating any errors and falling back to less strict parsing.

Things to parse and order of priority:
- inputs
  - input_messages (single part content)
  - input.value
- outputs
  - output_messages (single part)
  - output.value
- invocation parameters
-  prompt template variables
- tools

### PR DESCRIPTION
**Pull request by Parker-Stafford**
resolves #4922

- accumulates parsing errors while transforming a span to a playground instance and displays them in the playground
- displays output messages in the playground if present on the span

### Errors
image

### Output

Screenshot 2024-10-09 at 5 09 10 PM

### CODE REVIEW

File: app/src/pages/playground/playgroundUtils.ts [hunk 1]
```diff
@@ -1,20 +1,28 @@
...
+  outputSchema,
+} from "./schemas";
 import { PlaygroundSpan } from "./spanPlaygroundPageLoader";
 
 /**
  * Checks if a string is a valid chat message role
  */
 export function isChatMessageRole(role: unknown): role is ChatMessageRole {
-  return chatMessageRoles.includes(role as ChatMessageRole);
+  return chatMessageRolesSchema.safeParse(role).success;
```
Comment (Parker-Stafford) [inline app/src/pages/playground/playgroundUtils.ts:25]: we don't have to use this but i wanted to add it to the shcmea for ChatMessages so figured would update here as well

File: app/src/store/playgroundStore.tsx [hunk 2]
```diff
@@ -51,12 +51,12 @@ export type PlaygroundTemplate =
 /**
  * Array of roles for a chat message with a LLM
  */
-export const chatMessageRoles = ["user", "ai", "system", "tool"] as const;
-
-/**
- * The role of a chat message with a LLM
- */
-export type ChatMessageRole = (typeof chatMessageRoles)[number];
+export enum ChatMessageRole {
```
Comment (Parker-Stafford) [inline app/src/store/playgroundStore.tsx:54]: needed these actual values in a few places so switched to enum

File: app/src/pages/playground/PlaygroundOutput.tsx [hunk 3]
```diff
@@ -15,25 +20,61 @@ import {
   PlaygroundOutputSubscription$data,
   PlaygroundOutputSubscription$variables,
 } from "./__generated__/PlaygroundOutputSubscription.graphql";
+import { isChatMessages } from "./playgroundUtils";
 import { PlaygroundInstanceProps } from "./types";
 
 interface PlaygroundOutputProps extends PlaygroundInstanceProps {}
 
+export function PlaygroundOutputMessage({ message }: { message: ChatMessage }) {
```
Comment (Parker-Stafford) [inline app/src/pages/playground/PlaygroundOutput.tsx:28]: not sure if we want to show this card in all cases but wanted it for span parsing, seems to be a common way to display

File: app/src/pages/playground/PlaygroundOutput.tsx [hunk 4]
```diff
@@ -146,5 +189,13 @@ function PlaygroundOutputText(props: PlaygroundInstanceProps) {
...
     );
   }
-  return <span>{output}</span>;
+  return (
+    <PlaygroundOutputMessage
+      message={{
+        id: generateMessageId(),
+        content: output,
+        role: ChatMessageRole.ai,
+      }}
```
Comment (Parker-Stafford) [inline app/src/pages/playground/PlaygroundOutput.tsx:198]: can update output messages to not need id, seems fine to keep for now

File: app/src/pages/playground/__tests__/playgroundUtils.test.ts [hunk 5]
```diff
@@ -103,9 +198,9 @@ describe("getChatRole", () => {
 
   it("should return the ChatMessageRole if the role is included in ChatRoleMap", () => {
     expect(getChatRole("assistant")).toEqual("ai");
-    // expect(getChatRole("bot")).toEqual("ai");
-    // expect(getChatRole("system")).toEqual("system");
-    // expect(getChatRole("human:")).toEqual("user");
+    expect(getChatRole("bot")).toEqual("ai");
```
Comment (Parker-Stafford) [inline app/src/pages/playground/__tests__/playgroundUtils.test.ts:201]: oops

File: app/src/pages/playground/PlaygroundInstance.tsx [hunk 6]
```diff
@@ -28,6 +29,7 @@ export function PlaygroundInstance(props: PlaygroundInstanceProps) {
   return (
     <PanelGroup direction={isSingleInstance ? "horizontal" : "vertical"}>
       <Panel defaultSize={50} order={1} css={panelContentCSS}>
+        <PlaygroundInstanceErrors {...props} />
```
Comment (mikeldking) [inline app/src/pages/playground/PlaygroundInstance.tsx:32]: I think we eventually put this in a banner as a single "warning". Can always refactor

File: app/src/pages/playground/PlaygroundInstanceErrors.tsx [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+                  parsingErrors: parsingErrors.filter((_, j) => i !== j),
+                },
+              });
+            }}
+          >
+            {error}
+          </Alert>
+        );
+      })}
+    </Flex>
```
Comment (mikeldking) [inline app/src/pages/playground/PlaygroundInstanceErrors.tsx:46]: Maybe one alert with a `detail` element might be cleaner.

File: app/src/pages/playground/PlaygroundOutput.tsx (hunk 3 above)
Comment (mikeldking) [reply app/src/pages/playground/PlaygroundOutput.tsx:28]: I wouldn't export.

What do you mean you wanted it for "span parsing" ?

File: app/src/pages/playground/PlaygroundOutput.tsx [hunk 8]
```diff
@@ -15,25 +20,61 @@ import {
...
+        <PlaygroundOutputText key={runId} playgroundInstanceId={instanceId} />
+      );
+    }
+    if (isChatMessages(instance.output)) {
+      const messages = instance.output;
+
+      return messages.map((message, index) => {
+        return <PlaygroundOutputMessage key={index} message={message} />;
+      });
+    }
```
Comment (mikeldking) [inline app/src/pages/playground/PlaygroundOutput.tsx:60]: Oh as in this was the original output?

File: app/src/store/playgroundStore.tsx [hunk 9]
```diff
@@ -130,12 +130,19 @@ export interface PlaygroundInstance {
...
    * Whether or not the playground instance is actively running or not
    **/
   isRunning: boolean;
+  /**
+   * Errors that occurred during parsing of initial playground data.
+   * For example, when coming from a span to the playground, the span may
+   * not have the correct attributes, or the attributes may be of the wrong shape.
+   * This field is used to store any issues encountered when parsing to display in the playground.
+   */
+  parsingErrors?: string[];
```
Comment (mikeldking) [inline app/src/store/playgroundStore.tsx:145]: I'm not sure we need to store this in state - can't we just pass it as props in the `SpanPlayground` and display them at the top of the page if needed?

File: app/src/pages/playground/PlaygroundOutput.tsx (hunk 3 above)
Comment (Parker-Stafford) [reply app/src/pages/playground/PlaygroundOutput.tsx:28]: oh yeah export was a mistake

also the parsing wording was a brain fart haha, wanted it for displaying output messages parsed from the original span** is what i meant to say

File: app/src/pages/playground/PlaygroundInstanceErrors.tsx (hunk 7 above)
Comment (Parker-Stafford) [reply app/src/pages/playground/PlaygroundInstanceErrors.tsx:46]: yeah i like that, i can refactor

File: app/src/pages/playground/PlaygroundOutput.tsx (hunk 8 above)
Comment (Parker-Stafford) [reply app/src/pages/playground/PlaygroundOutput.tsx:60]: yes, displaying the original output but then also displaying new outputs in the same way

File: app/src/store/playgroundStore.tsx (hunk 9 above)
Comment (Parker-Stafford) [reply app/src/store/playgroundStore.tsx:145]: makes sense to me

**Review by mikeldking — APPROVED** [review]

**Pull request merged**